### PR TITLE
[ONNX] Fix controlflow shape inference with contrib op (#60707)

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63249 Enhance shape (#60335)
* #62765 [ONNX] Suppport torch.dot and torch.nn.utils.spectral_norm (#62596)
* #62764 [ONNX] Update repeat_interleave for dynamic repeats (#59979)
* #62763 [ONNX] Fix an issue that optimizations might adjust graph inputs unexpectedly. (#61280)
* **#62762 [ONNX] Fix controlflow shape inference with contrib op (#60707)**
* #62760 [ONNX] Fix for batchnorm training op mode (#52758)
* #62759 [ONNX] Remove aten parameter (#61652)
* #62758 [ONNX] Add support for opset14 in PT-ONNX exporter (#59486)
* #62757 [ONNX] Support lstm_cell symbolic (#61476)

`ONNXShapeTypeInference` for node `n` is skipped if `n` is non ONNX namespace, or if `n` contains any non ONNX namespace nodes. This prevents controlflow nodes containing contrib ops from running `SpecialPostProcess`, which sets up correct node output shape/type information in rare cases.

This PR depends on opset 14 export https://github.com/pytorch/pytorch/pull/59486

Co-authored-by: BowenBao <bowbao@microsoft.com>

Differential Revision: [D30375180](https://our.internmc.facebook.com/intern/diff/D30375180)

### CODE REVIEW

**Comment by msaroufim** [comment]
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30375180).

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 1]
```diff
@@ -401,5 +461,36 @@ std::vector<Value*> FixupONNXControlflowNode(Node* n, int opset_version) {
   }
 }
 
+void FixupONNXControlflowNodeOutputs(Node* n) {
+  switch (n->kind()) {
+    case ::c10::onnx::Loop: {
+      for (size_t i = 0; i < n->outputs().size(); ++i) {
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:467]: Please use `c10::irange`
```suggestion
      for (auto i : c10::irange(n->outputs().size())) {
```

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 2]
```diff
@@ -401,5 +461,36 @@ std::vector<Value*> FixupONNXControlflowNode(Node* n, int opset_version) {
...
+        auto loop_carried_output_size = n->blocks().at(0)->inputs().size() - 2;
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:468]: Why this needs to be computed in a loop?

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 3]
```diff
@@ -401,5 +461,36 @@ std::vector<Value*> FixupONNXControlflowNode(Node* n, int opset_version) {
...
+        if (i < loop_carried_output_size) {
+          n->output(i)->setType(n->blocks().at(0)->outputs().at(i + 1)->type());
+        } else {
+          auto type = n->blocks().at(0)->outputs().at(i + 1)->type();
```
Comment (malfet) [inline torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:472]: Nit
```suggestion
        auto type = n->blocks().at(0)->outputs().at(i + 1)->type();
        if (i < loop_carried_output_size) {
          n->output(i)->setType(type);
        } else {
```

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp]: thx, updated.

**Comment by msaroufim** [comment]
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30375180).
